README.txt
ISTOPICS
========

Configuration
-------------
The web server should have support for MySQL and a php interpreter with the MySQLi extension installed.

To allow for the .htaccess file to work, the server must be configured to allow this
    This can be accomplished on a Linux Apache server by editing /etc/apache2/sites-enabled/000-default.conf (or a very similar name) to include the following code:

    <Directory /var/www/html>
        AllowOverride All
    </Directory>

    Replace /var/www/html with the server's web root

    Then, the command "a2enmod rewrite" must be run to enable mod_rewrite
    Finally, restart the server: "service apache2 restart"

    If you want to use a virtual host, see your web server's documentation.

    Web root url should be localhost/istopics, or some such.

Installation
------------
-place the /istopics folder on the web server
    (probably at /var/www/html)
-run database_setup.sql in MySQL to set up the database and its tables as well as the mySQL user
-the mySQL user 'istopics'@'localhost' is created in the above file and granted necessary permissions
    update username and password in /istopics/db_credentials.php and in database_setup.sql
-to change a users's role from 'student' to 'admin', from MySQL run the command:
    UPDATE users SET role='admin' WHERE email='EMAIL OF THE USER TO BE UPDATATED';

To use Google Analytics, replace the x's on line 8 of istopics/analyticstranking.php with your own code

Auto Populate the Dabatase (testing)
------------------------------------
1. create a user profile to own all of the projects (from the website)

2. edit python scripts
   -edit db_filler/create_filler_user_project_connections.py by editing line 4 to hold the user id of the user created in step 1.
   -if you want to randomize the user id, uncomment lines 1 and 7
   -edit db_filler/create_filler_user_project_connections.py by editing line 8 to hold the last project id currently in the database (0 if having just run TRUNCATE on the projects table)

3. run db_filler/create_filler_projects.py from the same directory as db_filler/ISkeywords.csv
   -navigate to db_filler/ in the command line
   -enter "python create_filler_projects.py" (omitting the quotes) into the command line

4. run db_filler/create_filler_user-project-connections.py
   -navigate to db_filler/ in the command line
   -enter "python create_filler_user-project-connections.py" (omitting the quotes) into the command line

5. run the generated .sql files from MySQL
   -"SOURCE db_filler/filler_projects.sql;"
   -"SOURCE db_filler/filler_user_project_connections.sql;"
   -there will be some errors in the SQL, but that is okay. Enough INSERTs should succeed to give enough data to test.

Directory Layout
----------------
This project has the following directory structure:

istopics/ Contains meta information about the project and subdirectories containing parts of the project
|--db_filler/ Contains code to generate filler projects during development
|--istopics/ Contains PHP files
    |--css/ Contains custom CSS
    |--error/ Contains PHP pages for custom server error messages
    |--js/ Contains custom JavaScript code
|--majors/ Contains code to generate an HTML list of majors offered at The College of Wooster

Authors
-------
See AUTHORS.txt

Change Log
----------
See CHANGE_LOG.txt


CHANGE_LOG.txt
CHANGE LOG
==========
2017-01-27
-disabled table sorting because display was duplicating projects
-added licensing information to the footer
-added an about page

2017-01-10
-added ability to export projects to JSON or CSV

2017-01-09
-improved CSV exporter

2017-01-07
-added ability to set a custom major

2017-01-03
-can view project type and search/filter by project type

2017-01-02
-added table sorting

2016-12-27
-added multiple types of projects: Senior IS, Junior IS, and Other

2016-12-23
-added a simple table view of projects

2016-12-19
-preserve the favorite buttons when filtering projects
-added list of majors to advanced search

2016-12-01
-changed colors to represent Wooster

2016-11-03
-various bug fixes

2016-11-02
-can click on keywords and majors to show only projects with those keywords and majors

2016-10-26
-added basic version of advanced search

2016-10-19
-minor tweaks and bug fixes

2016-10-15
-improved backend for database operations
-can filter users in the admin interface
-minor bug fixes

2016-10-14
-improved sorting algorithm

2016-10-13
-added sorting of projects based on search terms

2016-10-06
-reduced number of iterations to go loop through projects

2016-05-04
-added ability to favorite projects
-updated styling

2016-04-29
-added dropdown navigation for small screens
-various styling improvements

2016-04-27
-improved who can see the Edit Profile button

2016-04-23
-added non-JS search

2016-04-20
-can delete user profiles
-can view a list of all users as an admin

2016-04-15
-moved profile diaplaying to a function
-added custom error pages

2016-04-13
-add required attribute to some some inputs
-added .htaccess to change urls
-admin pages can only be accessed from Wooster ips

2016-04-12
-can download all projects as an admin from adminInterface
-improved autocomplete
-improve project timestamp

2016-04-10
-added autocomplete of keywords when searching projects
-display when a project was last updated


2016-04-07
-ellipsify proposal and keywords on view all projects if they are too long
-fixed various display issues
-fixed various bugs

2016-04-06
-client side verification of wooster.edu email
-can save multiple majors
-throw error when making a duplicate user
-changed who can see edit project and edit user buttons
-added edit project button to user profile view
-can search through both accented and unaccented characters
-tell admin if user being promoted doesn't exist
-new user creation redirects to appropriate page

2016-04-04
-users can change passwords
-added student/faculty new user toggle
-modified views and forms to allow for faculty sign ins
-bug fixes

2016-04-03
-admins can promote other users to admin

2016-03-31
-store passwords as hashes
-add hint text to inputs
-more useful user creation error messages
-various bug fixes

2016-03-30
-touch up styling
-made alerts dismissible
-fixed OSx scrollbar issue

2016-03-26
-admin can update any project
-admin can update any user
-added links to user profiles from project views
-added link to admin interface
-created index page

2016-03-25
-added user roles
-added autofill for major selection
-user email must be unique
-various bug fixes

2016-03-24
-can reset database from admin interface

2016-03-21
-fixed project search counter
-added panels to project views
-started admin interface
-updated favicon.ico

2016-03-09
-changed abstract to proposal
-implemented comma-separated searching
-wrote scripts to generate filler projects

2016-03-08
-added searching for all project fields
-show number of search results

2016-03-05
-basic javaScript search functionality
-various bug fixes
-improved code readability
-moved javaScript to separate files
-auto generate user's email when they register

2016-03-02
-added ability to view user profiles
-added ability to update user profiles

2016-02-27
-added instructions for selecting multiple majors
-added keywords to showAllProjects.php
-moved database credentials to seperate file
-added more authentication and error/infomation messages

2016-02-26
-fixed bug that new users could not be created
-updated database_setup.sql

2016-02-25
-added author to project views
-added user authentication to update project page
-began some styling

2016-02-24
-can sign in
-linked projects to users

2016-02-23
-login no longer throws 500 error

2016-02-20
-added favicon.ico
-sanitize input
-added client side validation for projcet information
-started login capability

2016-02-18
-validation of user info input on both client and server
-can add users to database

2016-02-17
-can update a project
-can delete a project
-made output conform to w3c standards
-fixed various display issues

2016-02-16
-can view one project
-started search functionality

2016-02-13
-fixed view issues with view all projects
-added discipline column to database
-cleaned up new project form

2016-02-11
-can add projects from browser
-imporved look of pages

2016-02-10
-can view all projects (title and comments) from browser


istopics/dbToCSV.php
<?php
/**
 * Download projects from the database as a CSV file
 */

require_once 'db_credentials.php';
require_once 'checkSignIn.php';

if (issignedin() == 'admin') {

    $sql = "SELECT id, project_type, discipline, title, keywords, proposal, comments, date_created, last_updated FROM projects";

    if ((isset($_POST['discipline']) && $_POST['discipline'] != '') || $_POST['other_discipline'] != '' || $_POST['year'] != '' || $_POST['project_keywords'] != '' || (isset($_POST['project_type']) && $_POST['project_type'] != '')) {
        $sql .= " WHERE ";

        if ($_POST['other_discipline'] != '') {
            $sql .= "discipline LIKE '%{$_POST['other_discipline']}%'";
        }
        if (isset($_POST['discipline'])) {
            if ($_POST['other_discipline']) {
                $sql .= ' OR ';
            }
            if (! is_array($_POST['discipline'])) {
                $sql .= "discipline LIKE '%{$_POST['discipline']}%'";
            }
            elseif (count($_POST['discipline']) == 1) {
                $sql .= "discipline LIKE '%{$_POST['discipline'][0]}%'";
            }
            else {
                $first_disc = true;
                foreach ($_POST['discipline'] as $disc) {
                    if ($first_disc !== true) {
                        $sql .= ' OR ';
                    }
                    else {
                        $first_disc = false;
                    }
                    $sql .= "discipline LIKE '%{$disc}%'";
                }
            }
        }
        if ($_POST['project_keywords']) {
            if (isset($_POST['discipline'])) {
                $sql .= ' OR ';
            }
            $keywords = explode(',', $_POST['project_keywords']);

            $first_key = true;
            foreach ($keywords as $k) {
                if ($first_key !== true) {
                    $sql .= ' OR ';
                }
                else {
                    $first_key = false;
                }
                $sql .= "keywords LIKE '%{$k}%'";
            }
        }
        if (isset($_POST['project_type']) && count($_POST['project_type']) > 0) {
            if (isset($_POST['discipline']) || $_POST['other_discipline'] || $_POST['project_keywords']) {
                $sql .= ' AND ';
            }
            $sql .= '(';

            $first_type = true;
            foreach($_POST['project_type'] as $type) {
                if ($first_type !== true) {
                    $sql .= ' OR ';
                }
                else {
                    $first_type = false;
                }
                $sql .= "project_type LIKE '%{$type}%'";
            }

            $sql .= ')';
        }
        if (isset($_POST['year'])) {
            if (isset($_POST['discipline']) || $_POST['other_discipline'] || $_POST['project_keywords'] || isset($_POST['project_type'])) {
                $sql .= ' AND ';
            }
            $sql .= "YEAR(date_created)={$_POST['year']}";
        }
    }

    $result = $conn->query($sql);

    $outputFile = fopen("php://output", "w");

    if ($_POST['data_format'] == 'csv') {
        fputcsv($outputFile, array('id', 'project_type', 'major', 'title', 'keywords', 'proposal', 'comments', 'date_created', 'last_updated'));
    }
    else if ($_POST['data_format'] == 'json') {
        $projects = array();
    }

    // download the file
    $today = date("Y-m-d");
    header("Expires: Tue, 03 Jul 2001 06:00:00 GMT");
    header("Cache-Control: max-age=0, no-cache, must-revalidate, proxy-revalidate");

    // force download
    header("Content-Type: application/octet-stream");
    header("Content-Type: application/download");

    // output headers so that the file is downloaded rather than displayed
    header('Content-Type: application/csv; charset=utf-8');
    if ($_POST['data_format'] == 'csv') {
        header("Content-Disposition: attachment; filename=istopics_projects_{$today}.csv");
    }
    else if ($_POST['data_format'] == 'json') {
        header("Content-Disposition: attachment; filename=istopics_projects_{$today}.json");
    }

    // put results into the file
    if ($_POST['data_format'] == 'csv') {
        while ($row = $result->fetch_assoc()) {
            fputcsv($outputFile, $row);
        }
    }
    else if ($_POST['data_format'] == 'json') {
        while ($row = $result->fetch_assoc()) {
            $projects[] = array(
                'id' => $row['id'],
                'project_type' => $row['project_type'],
                'major' => $row['discipline'],
                'title' => $row['title'],
                'keywords' => $row['keywords'],
                'proposal' => $row['proposal'],
                'comments' => $row['comments'],
                'date_created' => $row['date_created'],
                'last_updated' => $row['last_updated'],
            );
        }

        $projects_file['projects'] = $projects;
        fwrite($outputFile, json_encode($projects_file));
    }

    fclose($outputFile);

    $conn->close();
}
else {
    $_SESSION["error"] = 1;
    $_SESSION["error_msg"] = "You are not authorized to perform this action.";

    header("Location: /istopics/project/all");
    exit();
}
